### DISCUSSION TITLE
Extracting single component from array variable.

### DISCUSSION THREAD

**Discussion by adityaghantasala** [idaholab/moose#20220] [Q&A General] [answered]
Hello Moose community !

How can I extract a single component of an array variable as a scalar variable ?
Use case : I would then make an object of a scalar kernel acting on this single component. 
Any pointers on how to achieve this will be helpful. 

Thank you in advance for the pointers.

**Comment by GiudGiud** [comment, marked as answer]
hello

Just to make sure, you want a single number, not a single component (a field then) of the array variable (composed of multiple fields, it s a vector field essentially)?

Guillaume

**Reply by adityaghantasala** [reply]
Hi Thank you for your reply. 
You mean a "number" like the variable number as here ?
https://github.com/idaholab/moose/blob/c98ae8b88fa3fc2dc27c667cef602fa437f11026/framework/src/systems/SystemBase.C#L825

Can this help creating a class(ideally a kernel) which can work on a single component(yes a field) ?

**Reply by GiudGiud** [reply]
ok so you need a field.
I meant a number as in: the value of the component 4 of variable array_with_5_components on element 12, which would have been 11.6 or something

This is the way to extract a component. However, this places it in a auxiliary variable. You cant create a 'kernel' in the sense of defining a non-linear equation (that will be solved by the nonlinear solver, not just substituting values in) for that component.
https://mooseframework.inl.gov/source/auxkernels/ArrayVariableComponent.html

You will be able to use this auxiliary variable to help define an equation for another variable, or compute other auxvariables and matprops

**Reply by adityaghantasala** [reply]
Hi @GiudGiud Thank you. That helps.
Another question. Is there a video or document explaining the software architecture or workflow of Moose framework ?

**Reply by GiudGiud** [reply]
Software architecture surely in the docs.
If you want the workflow, add this to your simulation, it ll give you a lot of info
```
[Debug]
  show_actions = true
[]
```

**Reply by GiudGiud** [reply]
Maybe this page will have some info you want?
https://mooseframework.inl.gov/framework_development/

**Reply by adityaghantasala** [reply]
> If you want the workflow, add this to your simulation, it ll give you a lot of info
> 
> ```
> [Debug]
>   show_actions = true
> []
> ```

@GiudGiud  Thank you for this pointer. This helps. But I did not get what you ment by `Software architecture surely in the docs` . You mean the doxygen pages with inheritance diagrams ?

I am looking more in the direction of which software design patterns are used, which classes fall under which pattern, which is the steering class for a given simulation and similar things which can give me a roadmap on how to see the dependency of particular class. Also this information is particularly helpful to debug some segfaults and errors which I get when experimenting with Moose. 

Thank you again for your help!

**Reply by GiudGiud** [reply]
The doc website is this one
https://mooseframework.inl.gov/index.html

There is also a lot of information about how classes are organized in the doxygen, and that will give you the dependencies of each class.

The steering class for a given simulation is usually the `Executioner`, though in fixed point iterations (multiapp-based setups) then it's the `FixedPointSolve`. 

We are actually reworking that system right now so this answer will change and be more clear in a few months.

**Reply by GiudGiud** [reply]
MOOSE papers are probably the best places to find data about the workflow actually. Unless there is an INL EXT report about it?
@permcody

**Reply by adityaghantasala** [reply]
> We are actually reworking that system right now so this answer will change and be more clear in a few months.

Interesting, can I follow this development ? is there a branch or a fork where this work is pushed to ?

> The steering class for a given simulation is usually the `Executioner`, though in fixed point iterations (multiapp-based setups) then it's the `FixedPointSolve`.

Thank you. I learnt this a couple of days ago, your previous pointers helped.
